amf-shapes/shared/src/main/scala/amf/shapes/internal/spec/jsonschema/parser/document/JsonSchemaLinker.scala

```diff
@@ -18,29 +18,40 @@ import amf.shapes.internal.validation.definitions.ShapeParserSideValidations.{
 import org.yaml.model.YNode
 
 trait NameExtraction {
-  def extract(ref: String): Either[String, String]
+  def extract(ref: String, declarationKey: Option[String]): Either[String, String]
 }
 
 trait JsonSchemaRefNameExtraction extends NameExtraction {
   protected def extractShapeName(uriFragment: String, definitionsKey: String): Option[String] = {
     val regex = ("^\\/" + definitionsKey + "\\/(?<shape>[^\\/]+)$").r
     regex.findFirstMatchIn(uriFragment).map(_.group(1))
   }
+
+  protected def validateDeclarationKey(
+      ref: String,
+      declarationKey: Option[String],
+      shape: String
+  ): Either[String, String] = declarationKey match {
+    case Some(dk) if ref.contains(dk) => Right(shape)
+    case Some(dk)                     => Left(s"The definition key present in the ref must be '$dk'")
+    case None                         => Right(shape)
+  }
 }
 
 object Draft2019NameExtraction extends JsonSchemaRefNameExtraction {
-  override def extract(ref: String): Either[String, String] = extractShapeName(ref, "\\$defs")
-    .orElse(extractShapeName(ref, "definitions")) match {
-    case Some(some) => Right(some)
-    case None       => Left(s"uriFragment '$ref' must be in the format of '#/definitions/<name>' or '#/$$defs/<name>'")
-  }
+  override def extract(ref: String, declarationKey: Option[String]): Either[String, String] =
+    extractShapeName(ref, "\\$defs").orElse(extractShapeName(ref, "definitions")) match {
+      case Some(some) => validateDeclarationKey(ref, declarationKey, some)
```

Comment:
For some reason you don't have formatting enabled.

Reply:
? I have formatting enabled